### PR TITLE
mon/AuthMonitor: speed up caps updates

### PR DESCRIPTION
**Pull request by jdurgin**
When trying to restrict access as much as possible for rbd, a viable approach is using a per-host key with access to only the namespaces of the images it is currently accessing. This implies updating these capabilities when images in use change. Doing this with the current 'ceph auth caps' command is quite slow, reaching 2 or fewer updates / second with 300 namespaces and 3 monitors running on tmpfs on a development machine. A major bottleneck in this case is the OSDCap parsing - it is quite cpu intensive, and performed with each valid_caps() call. Removing this speeds up the same test case by ~3x. Add an option to turn this off, but leave it on by default for usability.

There is likely room for improvement with rocksdb tuning as well, since it shows compaction latencies > 1s in these benchmarks on tmpfs, however that is left to future work.

### CODE REVIEW

**Review by jecluis — DISMISSED** [review]
I'm having trouble enabling this by default for all CLI commands, but otherwise looks good.

File: src/common/options.cc [hunk 1]
```diff
@@ -2150,6 +2150,14 @@ std::vector<Option> get_global_options() {
     .add_service("mon")
     .set_description("Timeout (in seconds) for smarctl to run, default is set to 5"),
 
+    Option("mon_auth_validate_all_caps", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
+    .set_default(true)
```
[the remark this answers is not in the document]
Comment (jecluis) [reply src/common/options.cc:2154]: Alternatively, and if this is not meant to be used solely by automation tools, an option `--do-not-validate` to the commands could also be a possibility?

**Comment by dillaman** [comment]
Any chance you saw any specific areas for improvement in the caps processing? I ask since the OSDs would also need to read/validate the caps for each connection, correct? If there is an obvious improvement we can make to cap processing when you have hundreds/thousands of individual OSD caps, it might be worth the quick investigation.

File: src/common/options.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jdurgin) [reply src/common/options.cc:2154]: I agree, this is why I left the default behavior - true means keep validating like we do today.

**Review by tchaikov — COMMENTED** [review]
> @dillaman there was a mess of boost::spirit shown by perf - I suspect using a different parsing framework might be necessary to avoid the generation of tons of intermediate objects, but that's mostly speculation.

probably to use a different format for representing the cap. like compiled/binary format crush versus the text format of crush.

File: src/common/legacy_config_opts.h [hunk 2]
```diff
@@ -305,6 +305,7 @@ OPTION(mon_keyvaluedb, OPT_STR)   // type of keyvaluedb backend
 OPTION(mon_debug_unsafe_allow_tier_with_nonempty_snaps, OPT_BOOL)
 OPTION(mon_osd_blocklist_default_expire, OPT_DOUBLE) // default one hour
 OPTION(mon_osd_crush_smoke_test, OPT_BOOL)
+OPTION(mon_auth_validate_all_caps, OPT_BOOL)
```
Comment (batrick) [inline src/common/legacy_config_opts.h:308]: We don't add new options to this file anymore?

File: src/mon/AuthMonitor.cc [hunk 3]
```diff
@@ -1276,7 +1276,14 @@ bool AuthMonitor::valid_caps(
     if (!moncap.parse(caps, out)) {
       return false;
     }
-  } else if (type == "mgr") {
+    return true;
+  }
+
+  if (!g_conf()->mon_auth_validate_all_caps) {
```
Comment (batrick) [inline src/mon/AuthMonitor.cc:1282]:
```suggestion
  if (!g_conf().get_val<bool>("mon_auth_validate_all_caps")) {
```

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/mon/AuthMonitor.cc (hunk 3 above)
Comment (jdurgin) [reply src/mon/AuthMonitor.cc:1282]: using this legacy form is a shortcut to avoid the config lock, you're right it's not needed here though

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
@jdurgin `make check` fails

**Comment by batrick** [comment]
@jdurgin we intend to backport this, right? We'll need a tracker ticket.

**Comment by batrick** [comment]
Tested here: https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20200904.221904

There's a lot of unrelated teuthology noise right now causing a lot of failures. Otherwise, this PR looks okay to me.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-09-06_13:32:55-rados-wip-kefu-testing-2020-09-06-2017-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-09-07_01:57:21-rados-wip-kefu-testing-2020-09-06-2017-distro-basic-smithi

failures tracked by 

- https://tracker.ceph.com/issues/47306
- https://tracker.ceph.com/issues/47179

and i tested locally using

```sh
../qa/run-standalone.sh osd-backfill-prio.sh
```

it passed.

**Review by tchaikov — APPROVED** [review]

File: src/common/options.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/common/options.cc:2154]: @jecluis i think the existing behavior is not changed with this changeset. could you take another look?

**Pull request merged**